util/types/datum.go

```diff
@@ -939,6 +941,68 @@ func (d *Datum) convertToUint(sc *variable.StatementContext, target *FieldType)
 	return ret, nil
 }
 
+func (d *Datum) convertToMysqlTimestamp(sc *variable.StatementContext, target *FieldType) (Datum, error) {
+	var (
+		ret Datum
+		t   Time
+		err error
+	)
+	fsp := DefaultFsp
+	if target.Decimal != UnspecifiedLength {
+		fsp = target.Decimal
+	}
+	loc := sc.TimeZone
+	switch d.k {
+	case KindMysqlTime:
+		t = d.GetMysqlTime()
+		switch {
+		case t.TimeZone == nil:
+			t.TimeZone = loc
+		case t.TimeZone == time.UTC:
+			// Convert to session timezone
+			if !t.IsZero() {
+				raw, err := t.Time.GoTime(time.UTC)
+				if err != nil {
+					return ret, errors.Trace(err)
+				}
+				converted := raw.In(loc)
+				t.Time = FromGoTime(converted)
+			}
+			t.TimeZone = loc
+		case t.TimeZone == sc.TimeZone:
+			break
+		default:
+			return ret, errors.New("something wrong")
+		}
+		t, err = t.RoundFrac(fsp)
+		if err != nil {
+			ret.SetValue(t)
+			return ret, errors.Trace(err)
+		}
+	case KindMysqlDuration:
```

Comment:
When converting `MysqlTime` to `MysqlDuration`, should session timezone be considered?